File: frontend/apps/ios/Pantopus/Core/Design/Components/RedactionScrim.swift
//
//  RedactionScrim.swift
//  Pantopus
//
//  B1.3 — identity-preview primitive for A18.5 "View as".
//
//  An overlay you place over any field or section to show it's withheld
//  from the currently-previewed viewer. It blurs + dims the wrapped
//  content and floats a centred lock chip ("Hidden from public", caller-
//  supplied copy) in `appSurfaceSunken`. The `level` tunes how aggressive
//  the treatment is:
//
//    • .hidden  — fully withheld: heavy blur, opaque sunken wash, content
//                 dropped from the accessibility tree.
//    • .fuzzed  — obscured but shape-legible: medium blur + wash.
//    • .partial — a hint of redaction: light blur, faint wash (used when
//                 a coarse value still shows, e.g. "Maple Heights area").
//
//  Composes over arbitrary children, so the View-As render (B5.2) can wrap
//  whichever rows the privacy resolver marks hidden without bespoke markup.
//

import SwiftUI

/// How hard the scrim hides its content. Raw treatment values are
/// geometry/opacity, not on the design-token scale, so they live on the
/// enum rather than as `Spacing`/`Radii` tokens.
public enum RedactionLevel: String, Sendable, CaseIterable, Hashable {
    case hidden
    case fuzzed
    case partial

    /// Gaussian blur applied to the wrapped content (points).
    var blurRadius: CGFloat {
        switch self {
        case .hidden: 10
        case .fuzzed: 6
        case .partial: 3
        }
    }

    /// Opacity of the `appSurfaceSunken` wash drawn over the content.
    var scrimOpacity: Double {
        switch self {
        case .hidden: 0.6
        case .fuzzed: 0.4
        case .partial: 0.18
        }
    }

    /// Opacity the content is dimmed to beneath the wash.
    var contentOpacity: Double {
        switch self {
        case .hidden: 0.35
        case .fuzzed: 0.6
        case .partial: 0.85
        }
    }
}

/// Blur + wash + lock-chip overlay for withheld profile fields.
@MainActor
public struct RedactionScrim<Content: View>: View {
    private let level: RedactionLevel
    private let label: String
    private let showsChip: Bool
    private let content: Content

    /// - Parameters:
    ///   - level: How aggressively to hide the content.
    ///   - label: Lock-chip copy (e.g. "Hidden from public").
    ///   - showsChip: Hide the lock chip when false (wash only).
    ///   - content: The field/section being redacted.
    public init(
        level: RedactionLevel,
        label: String = "Hidden",
        showsChip: Bool = true,
        @ViewBuilder content: () -> Content
    ) {
        self.level = level
        self.label = label
        self.showsChip = showsChip
        self.content = content()
    }

    public var body: some View {
        ZStack {
            content
                .blur(radius: level.blurRadius)
                .opacity(level.contentOpacity)
                .accessibilityHidden(level == .hidden)

            Rectangle()
                .fill(Theme.Color.appSurfaceSunken)
                .opacity(level.scrimOpacity)
                .allowsHitTesting(false)

            if showsChip {
                lockChip
            }
        }
        .clipped()
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("redactionScrim")
    }

    private var lockChip: some View {
        HStack(spacing: Spacing.s1) {
            Icon(.lock, size: 12, color: Theme.Color.appTextSecondary)
            Text(label)
                .font(.system(size: 11.5, weight: .semibold))
                .foregroundStyle(Theme.Color.appTextSecondary)
        }
        .padding(.horizontal, Spacing.s3)
        .padding(.vertical, Spacing.s1)
        .background(
            Capsule(style: .continuous).fill(Theme.Color.appSurfaceSunken)
        )
        .overlay(
            Capsule(style: .continuous).strokeBorder(Theme.Color.appBorderStrong, lineWidth: 1)
        )
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label)
        .accessibilityIdentifier("redactionScrim_lockChip")
    }
}

// MARK: - Preview

#Preview("RedactionScrim") {
    func sampleField(_ title: String, _ value: String) -> some View {
        VStack(alignment: .leading, spacing: Spacing.s1) {
            Text(title)
                .pantopusTextStyle(.overline)
                .foregroundStyle(Theme.Color.appTextMuted)
            Text(value)
                .font(.system(size: 13.5, weight: .semibold))
                .foregroundStyle(Theme.Color.appText)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Spacing.s3)
        .background(Theme.Color.appSurface)
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous))
    }

    return VStack(spacing: Spacing.s4) {
        ForEach(RedactionLevel.allCases, id: \.self) { level in
            RedactionScrim(level: level, label: "Hidden from public") {
                sampleField("Contact", "(555) 010-2837")
            }
        }
    }
    .padding(Spacing.s4)
    .background(Theme.Color.appBg)
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/components/RedactionScrim.kt
@file:Suppress("MagicNumber", "LongMethod", "MatchingDeclarationName", "UnusedPrivateMember")

package app.pantopus.android.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.blur
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.clipToBounds
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * B1.3 — identity-preview primitive for A18.5 "View as".
 *
 * Mirrors `Core/Design/Components/RedactionScrim.swift`. An overlay placed
 * over any field/section to show it's withheld from the currently-previewed
 * viewer: it blurs + dims the wrapped content and floats a centred lock
 * chip (caller-supplied copy) in `appSurfaceSunken`. The [level] tunes how
 * aggressive the treatment is — [RedactionLevel.Hidden] also drops the
 * content from the accessibility tree.
 *
 * Composes over arbitrary children, so the View-As render (B5.2) can wrap
 * whichever rows the privacy resolver marks hidden without bespoke markup.
 */

/**
 * How hard the scrim hides its content. Raw treatment values are
 * geometry/opacity, not on the design-token scale, so they live on the
 * enum rather than as `Spacing`/`Radii` tokens.
 */
enum class RedactionLevel(
    val blurRadius: Dp,
    val scrimOpacity: Float,
    val contentOpacity: Float,
) {
    Hidden(blurRadius = 10.dp, scrimOpacity = 0.6f, contentOpacity = 0.35f),
    Fuzzed(blurRadius = 6.dp, scrimOpacity = 0.4f, contentOpacity = 0.6f),
    Partial(blurRadius = 3.dp, scrimOpacity = 0.18f, contentOpacity = 0.85f),
}

/**
 * Blur + wash + lock-chip overlay for withheld profile fields.
 *
 * @param level How aggressively to hide the content.
 * @param label Lock-chip copy (e.g. "Hidden from public").
 * @param showsChip Hide the lock chip when false (wash only).
 * @param content The field/section being redacted.
 */
@Composable
fun RedactionScrim(
    level: RedactionLevel,
    modifier: Modifier = Modifier,
    label: String = "Hidden",
    showsChip: Boolean = true,
    content: @Composable () -> Unit,
) {
    Box(
        modifier =
            modifier
                .clipToBounds()
                .testTag("redactionScrim"),
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier =
                Modifier
                    .blur(level.blurRadius)
                    .alpha(level.contentOpacity)
                    .then(if (level == RedactionLevel.Hidden) Modifier.clearAndSetSemantics {} else Modifier),
        ) {
            content()
        }

        Box(
            modifier =
                Modifier
                    .matchParentSize()
                    .background(PantopusColors.appSurfaceSunken.copy(alpha = level.scrimOpacity)),
        )

        if (showsChip) {
            LockChip(label = label)
        }
    }
}

@Composable
private fun LockChip(label: String) {
    val shape = RoundedCornerShape(Radii.pill)
    Row(
        modifier =
            Modifier
                .clip(shape)
                .background(PantopusColors.appSurfaceSunken)
                .border(1.dp, PantopusColors.appBorderStrong, shape)
                .padding(horizontal = Spacing.s3, vertical = Spacing.s1)
                .testTag("redactionScrim_lockChip")
                .clearAndSetSemantics { contentDescription = label },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        PantopusIconImage(
            icon = PantopusIcon.Lock,
            contentDescription = null,
            size = 12.dp,
            tint = PantopusColors.appTextSecondary,
        )
        Text(
            text = label,
            style = PantopusTextStyle.small.copy(fontWeight = FontWeight.SemiBold, fontSize = 11.5.sp),
            color = PantopusColors.appTextSecondary,
        )
    }
}

@Preview(showBackground = true, widthDp = 360)
@Composable
private fun RedactionScrimPreview() {
    Column(
        modifier =
            Modifier
                .background(PantopusColors.appBg)
                .padding(Spacing.s4),
        verticalArrangement = Arrangement.spacedBy(Spacing.s4),
    ) {
        RedactionLevel.entries.forEach { level ->
            RedactionScrim(level = level, label = "Hidden from public") {
                SampleFieldPreview()
            }
        }
    }
}

@Composable
private fun SampleFieldPreview() {
    Column(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .padding(Spacing.s3),
        verticalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        Text(text = "CONTACT", style = PantopusTextStyle.overline, color = PantopusColors.appTextMuted)
        Text(text = "(555) 010-2837", style = PantopusTextStyle.small, color = PantopusColors.appText)
    }
}
